Refactor Analytics REST handler

jpkrohling opened the pull request.

jpkrohling commented:
This is just a preview, to see if we are in agreement about the proposed general approach. If it looks OK, I'll do similar changes to the other methods.

A problem I found was that I couldn't get swagger to add the `Criteria` to the list of parameters required on the `POST` method. This is the diff I get when comparing the version generated by master and the version generated by this branch:

```
$ diff target/generated/rest-apm.adoc /tmp/rest-apm.adoc
629a630,641
> *Body*
> 
> [cols="^20,55,^25", options="header"]
> |=======================
> |Required|Description|Data Type
> 
> |Yes|query criteria|<<Criteria>>
> 
> |=======================
> 
> 
> 
```

I'll keep digging on how to get it fixed, but I wanted to share this code for a discussion before I continue with this approach.

jpkrohling commented on server/rest/src/main/java/org/hawkular/apm/server/rest/AnalyticsHandler.java:
```diff
@@ -287,65 +291,8 @@ public void getPropertyInfo(
-                    .entity(errors).type(APPLICATION_JSON_TYPE).build());
-        }
-
+    public void getPrincipalInfo(@Suspended final AsyncResponse response, @BeanParam PrincipalsGetRequest request) {
```
Note that I replaced all the individual parameters by a `BeanParam`. While this requires an extra class, I believe this provides a good level of organization for the code. Besides, it allows for a generalization, useful for having the same actual handler code for both HTTP methods.

objectiser commented:
Looks ok if the swagger issue could be fixed - although rather than having a PrinciplsRequest (with GET/POST variations), these parameters are common across most of the operations, so they could be generic. Unless I misunderstood, but looked like you were going to have these three classes/interface per operation?

jpkrohling commented:
Actually, I just realized that the payload for the POST is not `{"criteria":{"principal":"jdoe"}}`, but `{"principal":"jdoe"}`, so, it should be perfectly possible to use the same PrincipalsRequest for both POST and GET. The difference is that swagger will not say "Criteria" as the parameter, but the individual parameters. Is that a problem?

jpkrohling commented:
> Unless I misunderstood, but looked like you were going to have these three classes/interface per operation?

*Usually*, I prefer to have one bean per method, but those can easily be changed if the parameters diverge in the future.

objectiser commented:
Yes that is fine for the POST parameters.

jpkrohling commented:
Actually, I think it would cause problems: the POST version requires two of the list parameters to be actual lists (from what I see on the code), while the GET requires values to be joined by pipes. Besides, the Request needs to have *either* `@QueryParam`s or `@FormParam`s. I think it's best to start already with one for each method :)

objectiser commented:
Starting to wonder if supporting both methods is worth it - I think the UI primarily uses the POST variation.

What do you think? To simplify the code - would then only require a single bean?

pavolloffay commented:
@jpkrohling could you please refactor  `TraceHandler` `fragments/search` in this PR and possibly other REST handler with multiple methods for the same uri?

jpkrohling commented:
@pavolloffay , sure, the PR is opened only to gather feedback on the general approach :) If we agree on how to move forward, I'll apply similar changes to all methods.

@objectiser , I'm all for consistency and simplicity, so, I'm in favor of having a single method for single operations. As this is a query, REST semantics "demand" it to be GET instead of POST. Is this a problem from the UI standpoint?

pavolloffay commented:
I would also vote for keeping only @GET methods for getting data. From the JS code perspective it is easier to construct JSON and send it in a body.

@ammendonca what do you think about keeping only GET methods.

objectiser commented:
That is the issue - it is easier for the UI to maintain a local object with the filter criteria and just POST it. Simplifies the UI code.

objectiser commented:
Also it avoids the issue with the lists, for filtering properties and faults, which have to be encoded when doing a GET request - not nice.

jpkrohling commented:
> Also it avoids the issue with the lists, for filtering properties and faults, which have to be encoded when doing a GET request - not nice.

Not sure what you mean. There are two ways of doing GET requests with parameters: a "query" parameter with an URI-encoded JSON, or regular query parameters. For lists of regular query parameters, JAX-RS transforms the items into a list of more than one parameter of the same is sent. Something like: `?foo=bar&foo=baz` becomes a `List<String>` with `bar` and `baz` as values.

jpkrohling commented:
s/regular query parameters/regular URI-encoded query parameters/

The URI-encoding should happen no matter what, even if you are joining the values with pipes.

objectiser commented:
@jpkrohling I was referring to the UI needing to construct the piped representation for the property name/value/type tuple. But @ammendonca may not have a problem with this, in which case it would simplify things.

jpkrohling commented:
Alright, I was missing this piece of the puzzle :) I guess the decision is now on @ammendonca 's hands. Let me know what's the preferred approach and I'll continue this PR with the desired format as input.

objectiser commented:
Ok thanks

ammendonca commented:
I think it's best to keep the `POST` method, even if `GET` seems to make more sense at first, but if we think the criteria can grow quite long (eg: multiple inclusion/exclusion of properties with long names). Also if we need to send some sensitive data as criteria, it can never be encrypted as it would with POST (over HTTPS). And yes, it's easier for the UI to handle it as a JSON object :)

jpkrohling commented:
Playing the devil's advocate for a moment:

> I think it's best to keep the POST method, even if GET seems to make more sense at first, but if we think the criteria can grow quite long (eg: multiple inclusion/exclusion of properties with long names).

There's actually no limit to GET parameters. If we keep only `name=value` pairs, instead of JSON, it's even better. 

One disadvantage with POST is that one cannot share the results with a colleague. The only way to get to the results is sharing what the parameters are to get that result.

> Also if we need to send some sensitive data as criteria, it can never be encrypted as it would with POST (over HTTPS).

For HTTPS connections, GET parameters are encrypted on transit like POST parameters.

> And yes, it's easier for the UI to handle it as a JSON object :)

The JSON can be sent via GET as well, not only as post. Like `http://server:port/?q={"param1":"value1"}`

ammendonca commented:
True, but the cleanness of POST payload data versus the complexity (eg: for debugging) and possible encoding issues of the URL query parameters, etc, makes it more appealing for me as a UI developer. I think we need JSON to be present as there is some hierarchy to the criteria parameters, IIRC.

In terms of security, didn't knew they'd be encrypted on transit, good to know. Still there are some concerns, like pasting/sharing/bookmarking URLs with sensitive data, or server logs containing those. Not sure how sensitive this data can be, but I know we share users email addresses for instance..

If GET is used, we can adapt to it for sure, I'm not very strongly opposed to it, just think POST makes things simpler.

objectiser commented:
Had a quick chat with Alex and it seems like your suggestion of using a query parameter with the JSON representation of the Criteria object would be a good compromise, although it would need to be tested out to make sure there were no encoding issues in the brower.

I've seen the q parameter used in some services - is this a convention? or could we use 'criteria' as the parameter name? Not too worried either way, but 'criteria' then matches up with the type name.

jpkrohling commented:
> I've seen the q parameter used in some services - is this a convention?

Not sure if it's a formal convention, but it's quite usual to have a `query` parameter to be named `q`. 

> to be tested out to make sure there were no encoding issues in the browser

Absolutely. Before building the URL, the UI has to do a proper URI encoding. But it would have to do it anyway, no matter the format of the value.

jpkrohling commented:
This is now ready for a second initial review. This implements the changes we discussed, namely:

- JSON as query parameter
- Query parameter is named "criteria"
- Only GET methods, no POST, to be in compliance with REST practices
- Changed from `AsyncResponse` to a regular response. If we need async processing for higher throughput, we can optimize it later

Because of those changes, the following is worth mentioning:

- As it's a regular GET request, the content type is irrelevant.
- The JSON in a query parameter cannot be converted directly by JAX-RS, so, we have to map it manually (see the `toCriteria()` method)
- This will break the compatibility with current clients, but I believe this was expected :)
- Ditto for the Swagger

objectiser commented:
@jpkrohling Only comment is whether the more generic CriteriaQuery bean would be better than one per method?

jpkrohling commented:
`CriteriaQuery` shouldn't be in the PR, it seems I forgot to remove it from a previous iteration. Only `PrincipalsQuery` is there. The criteria is received via query parameter as string, and the JSON is parsed into a proper `Criteria`.

objectiser commented:
Yes,but wondering as most/all the methods are going to be taking the same parameter, whether the bean should just be generic rather than specific to each op (e.g. PrincipalsQuery)?

jpkrohling commented:
We could try :)

jpkrohling commented:
This PR has been updated to implement the discussed changes all over the Analytics part. @objectiser , @pavolloffay , could you please do another round of reviews? If it looks OK, I'd apply the same concepts to the other handlers.

jpkrohling commented on client/analytics-service-rest-client/src/main/java/org/hawkular/apm/analytics/service/rest/client/AnalyticsServiceRESTClient.java:
```diff
@@ -1242,107 +387,136 @@ public void storeFragmentCompletionTimes(String tenantId, List<CompletionTime> c
+                    }
+                } else {
+                    if (log.isLoggable(Level.FINEST)) {
+                        log.finest("Failed to get unbound endpoints: status=["
```
Make this message generic.

jpkrohling commented on client/analytics-service-rest-client/src/main/java/org/hawkular/apm/analytics/service/rest/client/AnalyticsServiceRESTClient.java:
```diff
@@ -1242,107 +387,136 @@ public void storeFragmentCompletionTimes(String tenantId, List<CompletionTime> c
-                            + connection.getResponseMessage());
-                }
+            } catch (Exception e) {
+                log.log(Level.SEVERE, "Failed to get unbound endpoints", e);
```
This as well.

jpkrohling commented on client/analytics-service-rest-client/src/main/java/org/hawkular/apm/analytics/service/rest/client/AnalyticsServiceRESTClient.java:
```diff
@@ -1242,107 +387,136 @@ public void storeFragmentCompletionTimes(String tenantId, List<CompletionTime> c
+    }
 
-                resp.append(new String(b, 0, len));
+    private <T> T doInConnectionContext(String tenantId, URL url, Function<HttpURLConnection, T> function) {
```
Think about a better name for this method.

jpkrohling commented on server/rest/src/main/java/org/hawkular/apm/server/rest/AnalyticsHandler.java:
```diff
@@ -89,1578 +77,289 @@
-        }
-
+    public Response getUnboundEndpoints(@BeanParam UnboundEndpointsRequest request) {
+        return returnOrCatch(() -> {
```
This `returnOrCatch` also deserves a better name...

jpkrohling replied:
Note that this `returnOrCatch` might (should?) disappear on HWKAPM-570

jpkrohling commented on server/rest/src/main/java/org/hawkular/apm/server/rest/AnalyticsHandler.java:
```diff
@@ -89,1578 +77,289 @@
-
+    public Response getBoundEndpoints(@BeanParam BoundEndpointsRequest request) {
+        return returnOrCatch(() -> {
+            log.tracef("Get bound endpoints: name [%s] start [%s] end [%s]", request.getName(), request.getStartTime(), request.getEndTime());
```
For the future, we might want to remove those logs, making the services as EJBs and using interceptors to do such calls. This cleans up the business code, and provides a standard logging for all business methods.

jpkrohling commented on server/rest/src/main/java/org/hawkular/apm/server/rest/AnalyticsHandler.java:
```diff
@@ -89,1578 +77,289 @@
-        }
-
+    public Response getPropertyInfo(@BeanParam CriteriaRequest request) {
+        return criteriaResponse(request, (criteria, tenant) -> analyticsService.getPropertyInfo(tenant, criteria));
```
Another bad name: `criteriaResponse`.

jpkrohling commented on server/rest/src/main/java/org/hawkular/apm/server/rest/entity/BaseRequest.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * @author Juraci Paixão Kröhling
+ */
+public class BaseRequest {
```
`s/BaseRequest/TenantRequest` ?

jpkrohling commented on server/rest/src/main/java/org/hawkular/apm/server/rest/entity/CriteriaRequest.java:
```diff
@@ -0,0 +1,44 @@
+
+    public Criteria toCriteria() throws IOException {
+        if (null == criteria || criteria.isEmpty()) {
+            return new Criteria();
```
Is this OK, or should it return `null`? If `criteria.isEmpty()` == `new Criteria()`, then this is correct.

objectiser replied:
Yes this is ok.

jpkrohling commented on tests/server/wildfly/src/test/java/org/hawkular/apm/tests/wildfly/AnalyticsServiceRESTTest.java:
```diff
@@ -133,6 +128,7 @@ public void initTest() {
 
     @Test
     public void testGetUnboundEndpoints() {
+        service.clear(null);
```
Remove this, this was not intended to leave in the PR.

objectiser commented on client/analytics-service-rest-client/src/main/java/org/hawkular/apm/analytics/service/rest/client/AnalyticsServiceRESTClient.java:
```diff
@@ -1121,80 +327,19 @@ protected boolean buildQueryString(StringBuilder builder, Criteria criteria) {
-
-        if (buildQueryString(builder, criteria)) {
-            builder.append('&');
+        if (criteria.parameters().isEmpty()) {
```
As we are passing criteria in json form, we could get rid of this parameter encoding approach - so instead it may be better to have an 'isDefined()' method on the Criteria class to determine if any of the fields have been set?

jpkrohling replied:
https://issues.jboss.org/browse/HWKAPM-599 -- I'd do it as part of another (smaller) commit, as this one will probably only get bigger :)

objectiser commented on server/rest/src/main/java/org/hawkular/apm/server/rest/AnalyticsHandler.java:
```diff
@@ -89,1578 +77,289 @@
-                    + " millisecond since epoch") @DefaultValue("0") @QueryParam("startTime") long startTime,
-            @ApiParam(required = false,
-                    value = "traces before this time, "
- 
```
Maybe 'handleRequest' instead?

objectiser commented:
@jpkrohling Great refactor! Only had problems reviewing the test class due to the interleaving with all the POST methods removed - but assume all tests are functionally the same?

jpkrohling commented:
I updated the PR to include the changes discussed on the PR. Basically, I changed the new methods receiving lambdas to be named like `withX`. This is a convention from the Ruby world and I don't remember seeing a convention for this in Java yet. So, `withConnection` means that the block (lambda) wold receive a connection as parameter.

jpkrohling commented:
@objectiser , I rebased and squashed the last two commits. If you are OK with this patch, would you merge it? I'll then continue with similar changes for other handlers.

objectiser commented:
@jpkrohling Will try to look at it today - want to do a full build/test. Could you remove the 'do not merge' label?

objectiser commented:
@jpkrohling Thanks!
